Alternative Interface

sheavalentine-mdsol opened the pull request:
This is getting safe and unsafe agents, where the unsafe agent has an Activerecordesque interface

added safe, unsafe, safe?, and representor methods to agent
extended base_client and http_client to include information about interface methods
renamed Representor to RepresentorAgent
created SafeRepresentorAgent which behaves like the old Representor
added interface to convert representor between these types
renames Transition to TransitionAgent
made TransitionAgent dispatch to the correct Representor type
Spec'd out interface
renamed Representor and Transition references to reflect naming convention

sheavalentine-mdsol committed:
This is getting safe and unsafe agents, where the unsafe agent has an Activerecordesque interface

added safe, unsafe, safe?, and representor methods to agent
extended base_client and http_client to include information about interface methods
renamed Representor to RepresentorAgent
created SafeRepresentorAgent which behaves like the old Representor
added interface to convert representor between these types
renames Transition to TransitionAgent
made TransitionAgent dispatch to the correct Representor type
Spec'd out interface
renamed Representor and Transition references to reflect naming convention

HonoreDB commented on lib/farscape/client/http_client.rb:
```diff
@@ -33,6 +33,15 @@ def invoke(options = {})
+      def methods
+        {
+          unsafe: ['PUT', 'DELETE'],
+          idempotent: ['POST', 'PATCH'], # http://tools.ietf.org/html/rfc5789
```
This looks switched with the above line.

HonoreDB commented on lib/farscape/representor.rb:
```diff
@@ -2,36 +2,80 @@
+      embedded[method]
+    end
+
+    def get_transition(method, *args, &block)
```
This looks like it'd be cleaner with `request_parameters = nil` or (`{}`) rather than `*args`, since you're never expecting there to be more than one.

sheavalentine-mdsol replied:
This part is actually just wrong.  Fixing.

HonoreDB commented on lib/farscape/client/base_client.rb:
```diff
@@ -2,6 +2,26 @@ module Farscape
+        }
+      end
+
+      def safe_methods
```
API nitpick: if you def `.safe_method?(m)`, etc. instead, I feel like this would be more future-proof. `@client.safe_method?(m)` looks more robust than `@client.safe_methods.include?(m)`

HonoreDB commented on lib/farscape/client/base_client.rb:
```diff
@@ -2,6 +2,26 @@ module Farscape
     # Client independent of protocol, only used for HTTP for now
     class BaseClient
+
+      def methods
```
I'd rather not override the useful built-in "methods" method. `def verbs` or `def protocol_methods`.

HonoreDB commented on lib/farscape/representor.rb:
```diff
@@ -2,36 +2,80 @@
+  end
+
+  class RepresentorAgent < SafeRepresentorAgent
+    def method_missing(method, *args, &block)
```
Need to override `respond_to?` or `respond_to_missing?` when overriding `method_missing`.

HonoreDB commented on lib/farscape/representor.rb:
```diff
@@ -2,36 +2,80 @@
+    def method_missing(method, *args, &block)
+      method = method.to_s
+
+      get_embedded(method) || get_transition(method, *args, &block) || get_attribute(method) || raise(NoMethodError)
```
Proper style here is to call `super` rather than `raise(NoMethodError)`, for multiple good reasons, but I guess this is a placeholder anyway while we decide on behavior.

sheavalentine-mdsol committed:
EVERYTHING IS BROKEN!
Hooray!

miwest929 commented:
@sheavalentine-mdsol What's the status of this PR?

sheavalentine-mdsol committed:
Added smart parameter matching

sheavalentine-mdsol committed:
Merge branch 'brave_new_develop' into alternate_interface

Conflicts:
	spec/spec_helper.rb

sheavalentine-mdsol commented:
Once we track down the ActiveSupport issue it should be mergable.

HonoreDB commented:
When I tried bundle install locally, I got

```
Bundler could not find compatible versions for gem "activesupport":
  In Gemfile:
    moya (>= 0) ruby depends on
      rails (= 4.1.8) ruby depends on
        activesupport (= 4.1.8) ruby

    farscape (>= 0) ruby depends on
      activesupport (3.2.0)
```

removed the Gemfile.lock and it installed activesupport-4.1.8, and I got the same spec failure.

sheavalentine-mdsol committed:
We will progressively rely more and more on ActiveSupport

HonoreDB commented on lib/farscape/representor.rb:
```diff
@@ -2,36 +2,93 @@
     attr_reader :representor
 
     # TODO: Work with Representor to make this straight forwqrd.
+    # The conditional is to allow direct creation of a RepresentorAgent from a Representor
```
You've moved the conditional, best to move or remove this comment

HonoreDB commented on lib/farscape/representor.rb:
```diff
@@ -2,36 +2,93 @@
+
+  class RepresentorAgent < SafeRepresentorAgent
+    def method_missing(method, *args, &block)
+      begin
```
You can remove `begin` and the inner `end` from this method, since rescue blocks have method scope by default.

HonoreDB commented on lib/farscape/representor.rb:
```diff
@@ -2,36 +2,93 @@
+      begin
+        super
+      rescue NoMethodError => e
+        parmeters = args.first || {}
```
s/parmeters/parameters/g

HonoreDB commented on lib/farscape/transition.rb:
```diff
@@ -11,24 +12,36 @@ def initialize(transition, agent)
+
+    private
+
+    def match_params(args, options)
```
The Ruby-ish way to do this is something like

``` ruby
[:parameters, :attributes].each do |key_type|
  filtered_values = ...@transition.public_send(key_type)...
  options.public_send(:"#{key_type}=", filtered_values.merge...)
end
```

sheavalentine-mdsol replied:
That produces this code:

``` ruby
    def match_params(args, options)
      hash_filter = ->(hash,list) { hash.select { |k,_| list.include?(k) } }
      field_names = ->(field_list) { field_list.map { |field| field.name.to_sym } }
      [:parameters, :attributes].each do |key_type|
        filtered_values = hash_filter.call(args, field_names.call(@transition.public_send(key_type)))
        options.public_send(:"#{key_type}=", filtered_values.merge(options.public_send(key_type) || {}))
      end
      options
    end
```

While I like that it's more generic, I'm bothered by its aesthetics.

HonoreDB replied:
If you're grossed out by the repeated `public_send`s, I can't really blame you. The point for me was really to inline the lambdas, since they're now only called once:

``` ruby
def match_params(args, options)
      [:parameters, :attributes].each do |key_type|
        field_names = @transition.public_send(key_type).map { |field| field.name.to_sym }
        filtered_values = args.select { |k,_| field_names.include?(k) } }
        options.public_send(:"#{key_type}=", filtered_values.merge(options.public_send(key_type) || {}))
      end
      options
    end
```

sheavalentine-mdsol replied:
Done.

I like it.

miwest929 commented on lib/farscape/representor.rb:
```diff
@@ -2,36 +2,93 @@
-      Hash[representor.embedded.map do |k, reps| 
-           [k, reps.map { |rep| Representor.new(false, rep, agent) }]
-      end]
+      Hash[representor.embedded.map{ |k, reps| [k, reps.map { |rep| @agent.representor.new(false, rep, agent) }] }]
```
Please extract the inner `map` into it's own method for enhanced readability.

sheavalentine-mdsol committed:
broke up embedded inner map
refactored math_params trading aesthetics for generalizability
changed spelling of parameters
removed superfluous 'begin'
removed irrelevant comments

sheavalentine-mdsol committed:
removed lambdas, and prettified match_params

HonoreDB commented on spec/lib/farscape/cache_spec.rb:
```diff
@@ -11,10 +11,6 @@
     end
 
-    it 'does not require the full active_support suite' do
-      expect(Hash.new).not_to respond_to(:slice)
```
Can replace this assertion with `expect($".grep /active_support.rb/).not_to be_any`, which will only fail if we require the whole thing.

sheavalentine-mdsol replied:
Sure, so long as you can tell me the value of this test.

HonoreDB replied:
_crickets_

sheavalentine-mdsol commented:
I think all comments have been addressed.  Merge?

csavage-mdsol commented on spec/lib/farscape/integration/interface_spec.rb:
```diff
@@ -70,16 +70,76 @@
+          resources = agent.enter(entry_point)
+          expect(resources.drds).to be_a Farscape::RepresentorAgent
+        end
+        it "requires an ! on unsafe transitions" do
```
unsafe and idempotent
